src/Dialect/Krnl/KrnlOps.cpp

```diff
@@ -327,99 +326,106 @@ ParseResult KrnlIterateOp::parse(OpAsmParser &parser, OperationState &result) {
   if (parser.parseKeyword("with") || parser.parseLParen())
     return failure();
 
-  // A function to parse a lower or upper bound.
-  auto parseBound = [&result, &builder, &parser, &operandParser, &boundMaps](
-                        bool isUpper) -> ParseResult {
-    // 'min' / 'max' prefixes are generally syntactic sugar, but are required
-    // if the map has multiple results.
-    bool failedToParsedMinMax =
-        failed(parser.parseOptionalKeyword(isUpper ? "min" : "max"));
-
-    // Try parse an SSA operand.
-    if (succeeded(operandParser.ParseOptionalOperand(
-            builder.getIndexType(), result.operands))) {
-      AffineMap map = builder.getSymbolIdentityMap();
-      boundMaps.emplace_back(AffineMapAttr::get(map));
-      return success();
-    }
-
-    // Bound is not an SSA id, then it must be an integer.
-    // Parse an integer constant attribute.
-    // Get the attribute location.
-    llvm::SMLoc attrLoc = parser.getCurrentLocation();
-    Attribute boundAttr;
-    NamedAttrList tempBoundAttrContainer;
-    if (parser.parseAttribute(
-            boundAttr, builder.getIndexType(), "temp", tempBoundAttrContainer))
-      return failure();
+  if (failed(parser.parseOptionalRParen())) {
```

Comment:
Do you mind explaining why we define and use the `parseBound` only when failing the `parseOptionalRParen`.
Also, because the code is long, do you mind just labeling the end of the lambda function with a comment, as well as the code that follows as the continuation of when `parseOptionalRParen` fails... thanks

Reply:
parseBound was only needed when the parenthesis were non-empty

I realize that the code was rather convoluted so I refactored parseBound and parseLoop into a struct LoopParser that I hope makes it easier to follow